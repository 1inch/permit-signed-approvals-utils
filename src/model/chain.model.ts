export enum ChainId {
    etherumMainnet = 1,
    binanceMainnet = 56,
    polygonMainnet = 137,
}
