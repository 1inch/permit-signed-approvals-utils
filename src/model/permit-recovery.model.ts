export interface PermitRecoveryParams {
    chainId: number;
    tokenName: string;
    tokenAddress: string;
    callData: string;
    version?: string;
}

export interface SyncPermitRecoveryParams extends PermitRecoveryParams {
    nonce: number;
    isDomainWithoutVersion?: boolean;
}
