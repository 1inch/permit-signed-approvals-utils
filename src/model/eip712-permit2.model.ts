import { EIP712TypedData } from "./eip712.model";
import { TypedDataDomain, TypedDataField } from "@ethersproject/abstract-signer";
import { PermitBatchData, PermitSingle } from "@uniswap/permit2-sdk/dist/allowanceTransfer";

export type Eip712Permit2 = {
    types: Record<string, TypedDataField[]>;
    domain: TypedDataDomain;
    message: PermitSingle | PermitBatchData;
    primaryType: string;
} & EIP712TypedData;
