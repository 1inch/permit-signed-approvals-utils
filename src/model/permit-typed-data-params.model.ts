import {ChainId} from './chain.model';
import {EIP712Object, EIP712Parameter} from './eip712.model';

export interface PermitTypedDataParamsModel {
    chainId: ChainId;
    tokenName: string;
    tokenAddress: string;
    params: EIP712Object;
    isDomainWithoutVersion?: boolean;
    version?: string;
    permitModelFields?: EIP712Parameter[];
}
