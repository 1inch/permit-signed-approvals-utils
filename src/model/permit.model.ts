import {EIP712Object} from './eip712.model';

export interface PermitParams extends EIP712Object {
    owner: string;
    spender: string;
    value: string;
    nonce: number;
    deadline: number;
}

export interface DaiPermitParams extends EIP712Object {
    holder: string;
    spender: string;
    nonce: number;
    expiry: number;
    allowed: boolean;
}
