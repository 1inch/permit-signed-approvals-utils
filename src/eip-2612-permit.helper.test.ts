import {
    buildPermitTypedData,
    fromRpcSig,
    getPermitContractCallParams,
} from './eip-2612-permit.helper';
import {PermitParams} from './model/permit.model';

describe('eip-2612 permit helpers', () => {
    const chainId = 56;
    const contractAddress = '0x11111112542d85b3ef69ae05771c2dccff4faa26';
    const tokenAddress = '0x111111111117dc0aa78b770fa6a738034120c302';
    const tokenName = '1INCH Token';
    const walletAddress = '0x2c9b2dbdba8a9c969ac24153f5c1c23cb0e63914';

    const permitParams: PermitParams = {
        owner: walletAddress,
        spender: contractAddress,
        value: '1000000000',
        nonce: 0,
        deadline: 192689033,
    };

    it('buildPermitTypedData()', () => {
        const data = buildPermitTypedData({
            chainId,
            tokenName,
            tokenAddress,
            params: permitParams
        });

        expect(data).toMatchSnapshot();
    });

    it('fromRpcSig()', () => {
        const signature =
            '0x' +
            '3b448216a78f91e84db06cf54eb1e3758425bd97ffb9d6941ce437ec7a9c2c17' +
            '4c94f1fa492007dea3a3c305353bf3430b1ca506dd630ce1fd3da09bd387b2f3' +
            '1c';
        const data = fromRpcSig(signature);

        expect(data).toMatchSnapshot();
    });

    it('fromRpcSig() should throw error when signature length is invalid', () => {
        const signature =
            '0x' +
            '3b448216a78f91e84db06cf54eb1e3758425bd97ffb9d6941ce437ec7a9c2c17' +
            '4c94f1fa492007dea3a3c305353bf3430b1ca506dd630ce1fd3da09bd387b2f3' +
            '1c11001100';
        const t = () => fromRpcSig(signature);

        expect(t).toThrowError('Invalid signature length');
    });

    it('fromRpcSig() should cut 0x from the signature', () => {
        const signature =
            '3b448216a78f91e84db06cf54eb1e3758425bd97ffb9d6941ce437ec7a9c2c17' +
            '4c94f1fa492007dea3a3c305353bf3430b1ca506dd630ce1fd3da09bd387b2f3' +
            '1c';

        const without0x = fromRpcSig(signature);
        const with0x = fromRpcSig('0x' + signature);

        expect(without0x).toEqual(with0x);
    });

    it('fromRpcSig() should support both versions of `eth_sign` responses', () => {
        const signature =
            '3b448216a78f91e84db06cf54eb1e3758425bd97ffb9d6941ce437ec7a9c2c17' +
            '4c94f1fa492007dea3a3c305353bf3430b1ca506dd630ce1fd3da09bd387b2f3' +
            '1a';

        const result = fromRpcSig(signature);

        expect(result).toMatchSnapshot();
    });

    it('getPermitContractCallParams()', () => {
        const signature =
            '0x' +
            '3b448216a78f91e84db06cf54eb1e3758425bd97ffb9d6941ce437ec7a9c2c17' +
            '4c94f1fa492007dea3a3c305353bf3430b1ca506dd630ce1fd3da09bd387b2f3' +
            '1c';

        const result = getPermitContractCallParams(permitParams, signature);

        expect(result).toMatchSnapshot();
    });
});
