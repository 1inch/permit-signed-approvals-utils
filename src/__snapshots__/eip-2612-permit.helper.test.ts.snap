// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eip-2612 permit helpers buildPermitTypedData() 1`] = `
Object {
  "domain": Object {
    "chainId": 56,
    "name": "1INCH Token",
    "verifyingContract": "0x111111111117dc0aa78b770fa6a738034120c302",
    "version": "1",
  },
  "message": Object {
    "deadline": 192689033,
    "nonce": 0,
    "owner": "0x2c9b2dbdba8a9c969ac24153f5c1c23cb0e63914",
    "spender": "0x11111112542d85b3ef69ae05771c2dccff4faa26",
    "value": "1000000000",
  },
  "primaryType": "Permit",
  "types": Object {
    "EIP712Domain": Array [
      Object {
        "name": "name",
        "type": "string",
      },
      Object {
        "name": "version",
        "type": "string",
      },
      Object {
        "name": "chainId",
        "type": "uint256",
      },
      Object {
        "name": "verifyingContract",
        "type": "address",
      },
    ],
    "Permit": Array [
      Object {
        "name": "owner",
        "type": "address",
      },
      Object {
        "name": "spender",
        "type": "address",
      },
      Object {
        "name": "value",
        "type": "uint256",
      },
      Object {
        "name": "nonce",
        "type": "uint256",
      },
      Object {
        "name": "deadline",
        "type": "uint256",
      },
    ],
  },
}
`;

exports[`eip-2612 permit helpers fromRpcSig() 1`] = `
Object {
  "r": Object {
    "data": Array [
      59,
      68,
      130,
      22,
      167,
      143,
      145,
      232,
      77,
      176,
      108,
      245,
      78,
      177,
      227,
      117,
      132,
      37,
      189,
      151,
      255,
      185,
      214,
      148,
      28,
      228,
      55,
      236,
      122,
      156,
      44,
      23,
    ],
    "type": "Buffer",
  },
  "s": Object {
    "data": Array [
      76,
      148,
      241,
      250,
      73,
      32,
      7,
      222,
      163,
      163,
      195,
      5,
      53,
      59,
      243,
      67,
      11,
      28,
      165,
      6,
      221,
      99,
      12,
      225,
      253,
      61,
      160,
      155,
      211,
      135,
      178,
      243,
    ],
    "type": "Buffer",
  },
  "v": 28,
}
`;

exports[`eip-2612 permit helpers fromRpcSig() should support both versions of \`eth_sign\` responses 1`] = `
Object {
  "r": Object {
    "data": Array [
      59,
      68,
      130,
      22,
      167,
      143,
      145,
      232,
      77,
      176,
      108,
      245,
      78,
      177,
      227,
      117,
      132,
      37,
      189,
      151,
      255,
      185,
      214,
      148,
      28,
      228,
      55,
      236,
      122,
      156,
      44,
      23,
    ],
    "type": "Buffer",
  },
  "s": Object {
    "data": Array [
      76,
      148,
      241,
      250,
      73,
      32,
      7,
      222,
      163,
      163,
      195,
      5,
      53,
      59,
      243,
      67,
      11,
      28,
      165,
      6,
      221,
      99,
      12,
      225,
      253,
      61,
      160,
      155,
      211,
      135,
      178,
      243,
    ],
    "type": "Buffer",
  },
  "v": 53,
}
`;

exports[`eip-2612 permit helpers getPermitContractCallParams() 1`] = `
Array [
  "0x2c9b2dbdba8a9c969ac24153f5c1c23cb0e63914",
  "0x11111112542d85b3ef69ae05771c2dccff4faa26",
  "1000000000",
  192689033,
  28,
  Object {
    "data": Array [
      59,
      68,
      130,
      22,
      167,
      143,
      145,
      232,
      77,
      176,
      108,
      245,
      78,
      177,
      227,
      117,
      132,
      37,
      189,
      151,
      255,
      185,
      214,
      148,
      28,
      228,
      55,
      236,
      122,
      156,
      44,
      23,
    ],
    "type": "Buffer",
  },
  Object {
    "data": Array [
      76,
      148,
      241,
      250,
      73,
      32,
      7,
      222,
      163,
      163,
      195,
      5,
      53,
      59,
      243,
      67,
      11,
      28,
      165,
      6,
      221,
      99,
      12,
      225,
      253,
      61,
      160,
      155,
      211,
      135,
      178,
      243,
    ],
    "type": "Buffer",
  },
]
`;
